File: lightbulb-dialogs/src/main/java/com/github/RooneyAndShadows/lightbulb/dialogs/dialog_custom/CustomDialogBuilder.java
package com.github.rooneyandshadows.lightbulb.dialogs.dialog_custom;

import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogBuilder;
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogFragment;

import androidx.fragment.app.FragmentManager;
import androidx.lifecycle.LifecycleOwner;

import static com.github.rooneyandshadows.lightbulb.dialogs.dialog_custom.CustomDialog.*;


public class CustomDialogBuilder<DialogType extends CustomDialog> extends BaseDialogBuilder<DialogType> {
    private final CustomDialogInitializer<DialogType> dialogInitializer;
    private final CustomDialogInflater dialogInflater;
    private boolean loading = false;

    public CustomDialogBuilder(LifecycleOwner lifecycleOwner, FragmentManager manager, String dialogTag, CustomDialogInitializer<DialogType> dialogInitializer, CustomDialogInflater dialogInflater) {
        super(lifecycleOwner, manager, dialogTag);
        this.dialogInitializer = dialogInitializer;
        this.dialogInflater = dialogInflater;
    }

    public CustomDialogBuilder(FragmentManager manager, String dialogTag, CustomDialogInitializer<DialogType> dialogInitializer, CustomDialogInflater dialogInflater) {
        super(manager, dialogTag);
        this.dialogInitializer = dialogInitializer;
        this.dialogInflater = dialogInflater;
    }

    @Override
    public CustomDialogBuilder<DialogType> withTitle(String title) {
        return (CustomDialogBuilder<DialogType>) super.withTitle(title);
    }

    @Override
    public CustomDialogBuilder<DialogType> withMessage(String message) {
        return (CustomDialogBuilder<DialogType>) super.withMessage(message);
    }

    @Override
    public CustomDialogBuilder<DialogType> withPositiveButton(DialogButtonConfiguration positiveButtonConfiguration, BaseDialogFragment.DialogButtonClickListener onClickListener) {
        return (CustomDialogBuilder<DialogType>) super.withPositiveButton(positiveButtonConfiguration, onClickListener);
    }

    @Override
    public CustomDialogBuilder<DialogType> withNegativeButton(DialogButtonConfiguration negativeButtonConfiguration, BaseDialogFragment.DialogButtonClickListener onClickListener) {
        return (CustomDialogBuilder<DialogType>) super.withNegativeButton(negativeButtonConfiguration, onClickListener);
    }

    @Override
    public CustomDialogBuilder<DialogType> withOnCancelListener(BaseDialogFragment.DialogCancelListener listener) {
        return (CustomDialogBuilder<DialogType>) super.withOnCancelListener(listener);
    }

    @Override
    public CustomDialogBuilder<DialogType> withOnShowListener(BaseDialogFragment.DialogShowListener listener) {
        return (CustomDialogBuilder<DialogType>) super.withOnShowListener(listener);
    }

    @Override
    public CustomDialogBuilder<DialogType> withOnHideListener(BaseDialogFragment.DialogHideListener listener) {
        return (CustomDialogBuilder<DialogType>) super.withOnHideListener(listener);
    }

    @Override
    public CustomDialogBuilder<DialogType> withCancelOnClickOutsude(boolean closeOnClickOutside) {
        return (CustomDialogBuilder<DialogType>) super.withCancelOnClickOutsude(closeOnClickOutside);
    }

    @Override
    public CustomDialogBuilder<DialogType> withDialogType(DialogTypes dialogType) {
        return (CustomDialogBuilder<DialogType>) super.withDialogType(dialogType);
    }

    @Override
    public CustomDialogBuilder<DialogType> withAnimations(DialogAnimationTypes animation) {
        return (CustomDialogBuilder<DialogType>) super.withAnimations(animation);
    }

    @Override
    public CustomDialogBuilder<DialogType> withDialogCallbacks(DialogCallbacks callbacks) {
        return (CustomDialogBuilder<DialogType>) super.withDialogCallbacks(callbacks);
    }

    public CustomDialogBuilder<DialogType> withLoading(boolean isLoading) {
        this.loading = isLoading;
        return this;
    }

    public interface CustomDialogInitializer<DialogType extends CustomDialog> {
        DialogType initialize(
                String title,
                String message,
                DialogButtonConfiguration positiveButtonConfiguration,
                DialogButtonConfiguration negativeButtonConfiguration,
                boolean cancelable,
                boolean loading,
                DialogTypes dialogType,
                DialogAnimationTypes animationType
        );
    }

    @SuppressWarnings("unchecked")
    @Override
    public DialogType buildDialog() {
        DialogType dialogFragment = (DialogType) fragmentManager.findFragmentByTag(dialogTag);
        if (dialogFragment == null)
            dialogFragment = dialogInitializer.initialize(
                    title, message, positiveButtonConfiguration, negativeButtonConfiguration,
                    cancelableOnClickOutside, loading, dialogType, animation
            );
        dialogFragment.setLifecycleOwner(dialogLifecycleOwner);
        dialogFragment.setDialogInflater(dialogInflater);
        dialogFragment.setDialogCallbacks(dialogCallbacks);
        dialogFragment.setFragmentManager(fragmentManager);
        dialogFragment.setDialogTag(dialogTag);
        dialogFragment.addOnShowListener(onShowListener);
        dialogFragment.addOnPositiveClickListener(onPositiveClickListener);
        dialogFragment.addOnNegativeClickListeners(onNegativeClickListener);
        dialogFragment.addOnHideListener(onHideListener);
        dialogFragment.addOnCancelListener(onCancelListener);
        return dialogFragment;
    }
}

File: lightbulb-dialogs/src/main/java/com/github/rooneyandshadows/lightbulb/dialogs/dialog_custom/CustomDialogBuilder.kt
package com.github.rooneyandshadows.lightbulb.dialogs.dialog_custom

import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.fragment.app.FragmentManager
import androidx.lifecycle.LifecycleOwner
import com.github.rooneyandshadows.lightbulb.dialogs.base.BaseDialogBuilder
import com.github.rooneyandshadows.lightbulb.dialogs.base.internal.DialogAnimationTypes
import com.github.rooneyandshadows.lightbulb.dialogs.base.internal.DialogButton
import com.github.rooneyandshadows.lightbulb.dialogs.base.internal.DialogTypes
import com.github.rooneyandshadows.lightbulb.dialogs.base.internal.callbacks.*
import com.github.rooneyandshadows.lightbulb.dialogs.dialog_custom.CustomDialog.CustomDialogInflater

@Suppress("unused")
class CustomDialogBuilder<DialogType : CustomDialog> @JvmOverloads constructor(
    dialogTag: String,
    dialogParentFragmentManager: FragmentManager,
    private val dialogInitializer: CustomDialogInitializer<DialogType>,
    private val dialogInflater: CustomDialogInflater? = null,
    dialogLifecycleOwner: LifecycleOwner? = null,
    initialDialogState: Bundle? = null,
) : BaseDialogBuilder<DialogType>(dialogTag, dialogParentFragmentManager, dialogLifecycleOwner, initialDialogState) {
    private var loading = false

    @JvmOverloads
    constructor(
        dialogTag: String,
        fragment: Fragment,
        dialogInitializer: CustomDialogInitializer<DialogType>,
        dialogInflater: CustomDialogInflater? = null,
        initialDialogState: Bundle? = null
    ) : this(
        dialogTag,
        fragment.childFragmentManager,
        dialogInitializer,
        dialogInflater,
        fragment,
        initialDialogState
    )

    @JvmOverloads
    constructor(
        dialogTag: String,
        activity: FragmentActivity,
        dialogInitializer: CustomDialogInitializer<DialogType>,
        dialogInflater: CustomDialogInflater? = null,
        initialDialogState: Bundle? = null
    ) : this(
        dialogTag,
        activity.supportFragmentManager,
        dialogInitializer,
        dialogInflater,
        activity,
        initialDialogState
    )

    @Override
    override fun setupNonRetainableSettings(dialog: DialogType) {
        dialog.apply {
            setDialogInflater(dialogInflater)
        }
    }

    @Override
    override fun setupRetainableSettings(dialog: DialogType) {
        dialog.apply {
            setLoading(loading)
        }
    }

    @Override
    override fun initializeNewDialog(): DialogType {
        return dialogInitializer.initialize()
    }

    @Override
    override fun withTitle(title: String): CustomDialogBuilder<DialogType> {
        return super.withTitle(title) as CustomDialogBuilder<DialogType>
    }

    @Override
    override fun withMessage(message: String): CustomDialogBuilder<DialogType> {
        return super.withMessage(message) as CustomDialogBuilder<DialogType>
    }

    @Override
    override fun withButton(configuration: DialogButton): CustomDialogBuilder<DialogType> {
        return super.withButton(configuration) as CustomDialogBuilder<DialogType>
    }

    @Override
    override fun withOnCancelListener(listener: DialogCancelListener): CustomDialogBuilder<DialogType> {
        return super.withOnCancelListener(listener) as CustomDialogBuilder<DialogType>
    }

    @Override
    override fun withOnShowListener(listener: DialogShowListener): CustomDialogBuilder<DialogType> {
        return super.withOnShowListener(listener) as CustomDialogBuilder<DialogType>
    }

    @Override
    override fun withOnHideListener(listener: DialogHideListener): CustomDialogBuilder<DialogType> {
        return super.withOnHideListener(listener) as CustomDialogBuilder<DialogType>
    }

    @Override
    override fun withCancelOnClickOutside(closeOnClickOutside: Boolean): CustomDialogBuilder<DialogType> {
        return super.withCancelOnClickOutside(closeOnClickOutside) as CustomDialogBuilder<DialogType>
    }

    @Override
    override fun withDialogType(dialogType: DialogTypes): CustomDialogBuilder<DialogType> {
        return super.withDialogType(dialogType) as CustomDialogBuilder<DialogType>
    }

    @Override
    override fun withAnimations(animation: DialogAnimationTypes): CustomDialogBuilder<DialogType> {
        return super.withAnimations(animation) as CustomDialogBuilder<DialogType>
    }

    @Override
    override fun withDialogListeners(listeners: DialogListeners): CustomDialogBuilder<DialogType> {
        return super.withDialogListeners(listeners) as CustomDialogBuilder<DialogType>
    }

    fun withLoading(isLoading: Boolean): CustomDialogBuilder<DialogType> {
        loading = isLoading
        return this
    }

    fun interface CustomDialogInitializer<out DialogType : CustomDialog> {
        fun initialize(): DialogType
    }
}